crates/oci/src/client.rs

```diff
@@ -164,12 +163,25 @@ impl Client {
         locked.components = components;
         locked.metadata.remove("origin");
 
-        let oci_config = Config {
-            data: serde_json::to_vec(&locked)?,
-            media_type: SPIN_APPLICATION_MEDIA_TYPE.to_string(),
-            annotations: None,
+        // Push layer for locked spin application config
+        let locked_config_layer = ImageLayer::new(
+            serde_json::to_vec(&locked).context("could not serialize locked config")?,
+            SPIN_APPLICATION_MEDIA_TYPE.to_string(),
+            None,
+        );
+        layers.push(locked_config_layer);
+
+        // Construct empty/default OCI config file. Data may be parsed according to
+        // the expected config structure per the image spec, so we want to ensure it conforms.
+        // (See https://github.com/opencontainers/image-spec/blob/main/config.md)
+        // TODO: Explore adding data applicable to a Spin app.
+        let oci_config_file = ConfigFile {
```

Comment:
We should also set the `OS` and `ARCH`, on the image. Because we are moving away from using the `ArtifactType` (see discussions in the [spec repo](https://github.com/opencontainers/image-spec/pull/1137#issuecomment-1767144964)), we need a way to identify this image.  These values are [recommended ](https://github.com/opencontainers/image-spec/blob/main/config.md#properties) to be set to Go Language values.  

> Configurations SHOULD use, and implementations SHOULD understand, values listed in the Go Language document for [GOARCH](https://golang.org/doc/install/source#environment).

https://github.com/containerd/runwasi/blob/828f1cce03feb6a3b2d894c27624e7d7c5cef0c2/crates/wasi-demo-app/build.rs#L40-L43

In Runwasi, to start, we will be looking for `ARCH=wasm` in order to know we should read and pass the custom layers to the runtime.  We will read `OS` part and pass it to the runtime so it can be used to determine whether to use modules/components.

Reply:
In the linked runwasi code, I see "wasi" is used as OS, though the official valid combo from https://go.dev/doc/install/source#environment appears to be "wasip1" (w/ ARCH of `wasm`).  I gather for containerd's needs, "wasi" is okay?  I assume "wasip2" support (both in go and containerd/etc) is in the near future?

Reply:
The image was originally written before golang released the support for wasi and set the OS/ARCH values so we didn't know what it would be.  

As of right now, the OS flag isn't going to be used by runwasi/containerd but passed to the runtime to use to make discussions so TBH it doesn't matter _yet_ since runwasi/containerd are not triggering off it (for now).  That said, the code linked will be updated to be wasip1 (have pr open for it) since the module being built there is actually wasip1 via rust. With the wasmtime/wasmedge shims we will use this to determine if it is module or component and set up the runtime appropriately.

In the future (once things stabilize more) the OS value and possibly `os_features` will become more important as we hope to have more consistency across the eco system and to have containerd do more based on that information.  Right now, containerd simply passes this info onto the shim (runwasi).  I am working on drafting up a doc that will point to how this looks in the future and will share with you folks once ready.

Does Spin adhere to the P1 or have a custom format?

Reply:
> Does Spin adhere to the P1 or have a custom format?

Well, that's a bit tricky.  The `.wasm` file(s) included in a Spin app may be preview 1 or may even be preview 2 depending on the language-specific build toolchain used.  Most likely preview 1 at this point.  But then, the higher-level Spin application itself, which this OCI image represents (eg consisting of one or more p1 or p2 `.wasm` files plus zero or more data files/layers and the locked spin config layer), would be considered a 'custom' format and require the Spin runtime (or conformant runtime) to run successfully.

Reply:
Since there is a mix, I think either wasip1 or wasip2 would make most sense.  And this is an example where I think we have a little time before everything stabilizes.  At that point I would think there would be "everything is a component" as mentioned when discussing layer types https://github.com/fermyon/spin/issues/1857#issuecomment-1755925324